packages/backend-function/src/factory.ts

```diff
@@ -260,6 +261,17 @@ class AmplifyFunction
     this.resources = {
       lambda: functionLambda,
     };
+
+    // This is a hack to generate a type definition file after synthesis
+    this.node.addValidation({
+      validate: (): string[] => {
+        new FunctionEnvironmentTypeGenerator(
+          id,
+          props.entry
+        ).generateTypeDefFile();
+        return [];
+      },
+    });
```

Comment:
Shouldn't this be tied to an `amplify generate ...` command rather than to synthesis? What'd be the DX like in current case?
1. Use `defineFunction` in your project
2. Run `amplify sandbox`
3. Come back to your project and use the environment variables?

Reply:
Yup that would be the DX for this.

If we tied this to just the `amplify generate ...` command then customers would have to run this after every change that adds/removes function env vars (once we start adding dynamic env vars) to benefit from the types. We can certainly add a `generate` command after but we discussed with Josef and decided having it tied to the end of synthesis for now.